File: cef-java-32/src/main/java/com/ui/cef_control/config/AppConfig.java
package com.anca.appl.fw.gui.cef_control.config;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 *
 */
public final class AppConfig {

	private final int ipcPort;
	private final String sessionToken;
	private final String startUrl;
	private final String windowId;

	private static final int MIN_PORT = 1024;
	private static final int MAX_PORT = 65535;
	private static final Set<String> REQUIRED_FLAGS = Set.of(
			"--ipcPort",
			"--sessionToken",
			"--startUrl",
			"--windowId"
	);

	private AppConfig(int ipcPort, String sessionToken, String startUrl, String windowId) {
		this.ipcPort = ipcPort;
		this.sessionToken = sessionToken;
		this.startUrl = startUrl;
		this.windowId = windowId;
	}

	public static AppConfig fromArgs(String[] args) {
		Map<String, String> parsed = parseArgs(args);
		validateAllRequiredPresent(parsed);

		int ipcPort = parseIpcPort(parsed.get("--ipcPort"));
		String sessionToken = validateNonEmpty(parsed.get("--sessionToken"), "sessionToken");
		String startUrl = validateNonEmpty(parsed.get("--startUrl"), "startUrl");
		String windowId = validateNonEmpty(parsed.get("--windowId"), "windowId");

		return new AppConfig(ipcPort, sessionToken, startUrl, windowId);
	}

	private static Map<String, String> parseArgs(String[] args) {
		Map<String, String> parsed = new HashMap<>();

		for (int i = 0; i < args.length; i++) {
			String arg = args[i];

			if (!arg.startsWith("--")) {
				throw new InvalidConfigException("Invalid argument format: " + arg);
			}

			if (i + 1 >= args.length) {
				throw new InvalidConfigException(
						"Flag " + arg + " requires a value"
				);
			}

			if (!REQUIRED_FLAGS.contains(arg)) {
				throw new InvalidConfigException("Unknown flag: " + arg);
			}

			parsed.put(arg, args[i + 1]);
			i++;
		}

		return parsed;
	}

	private static void validateAllRequiredPresent(Map<String, String> parsed) {
		for (String flag : REQUIRED_FLAGS) {
			if (!parsed.containsKey(flag)) {
				String paramName = flag.substring(2);
				throw new InvalidConfigException(
						"Required parameter missing: " + paramName
				);
			}
		}
	}

	private static int parseIpcPort(String value) {
		int port;
		try {
			port = Integer.parseInt(value);
		} catch (NumberFormatException e) {
			throw new InvalidConfigException(
					"ipcPort must be a valid integer",
					e
			);
		}

		if (port < MIN_PORT || port > MAX_PORT) {
			throw new InvalidConfigException(
					"ipcPort must be in range " + MIN_PORT + "-" + MAX_PORT +
							", got: " + port
			);
		}

		return port;
	}

	private static String validateNonEmpty(String value, String paramName) {
		if (value == null || value.isEmpty()) {
			throw new InvalidConfigException(
					paramName + " cannot be empty"
			);
		}
		return value;
	}

	public int getIpcPort() {
		return ipcPort;
	}

	public String getSessionToken() {
		return sessionToken;
	}

	public String getStartUrl() {
		return startUrl;
	}

	public String getWindowId() {
		return windowId;
	}
}

File: cef-parallel/src/core/AppConfig.cpp
#include "../../inc/core/AppConfig.h"
#include "../../inc/core/InvalidConfigException.h"
#include <stdexcept>
#include <sstream>
#include <algorithm>
#include <climits>

namespace cef_ui {
namespace core {

AppConfig::AppConfig(uint16_t ipc_port,
                     const std::string& session_token,
                     const std::string& start_url,
                     uint32_t window_id)
    : ipc_port_(ipc_port),
      session_token_(session_token),
      start_url_(start_url),
      window_id_(window_id) {}

AppConfig AppConfig::FromArgs(const std::vector<std::string>& args) {
  if (args.empty()) {
    throw InvalidConfigException("ConfigError:  No arguments provided");
  }

  uint16_t ipc_port = 0;
  std::string session_token;
  std::string start_url;
  uint32_t window_id = 0;

  bool has_ipc_port = false;
  bool has_session_token = false;
  bool has_start_url = false;
  bool has_window_id = false;

  // PHASE 1: Parse all arguments WITHOUT validation (collect values only)
  for (size_t i = 0; i < args.size(); ++i) {
    const std::string& arg = args[i];

    if (arg == "--ipcPort") {
      if (i + 1 >= args.size()) {
        throw InvalidConfigException("ConfigError:  --ipcPort requires a value");
      }
      try {
        int port_val = std::stoi(args[i + 1]);
        ipc_port = static_cast<uint16_t>(port_val);
        has_ipc_port = true;
        ++i;  // Skip the value
      } catch (const std::invalid_argument&) {
        throw InvalidConfigException("ConfigError:  --ipcPort value is not a valid integer");
      } catch (const std::out_of_range&) {
        throw InvalidConfigException("ConfigError:  --ipcPort value is out of range");
      }
    } else if (arg == "--sessionToken") {
      if (i + 1 >= args.size()) {
        throw InvalidConfigException("ConfigError:  --sessionToken requires a value");
      }
      session_token = args[i + 1];
      has_session_token = true;
      ++i;  // Skip the value
    } else if (arg == "--startUrl") {
      if (i + 1 >= args.size()) {
        throw InvalidConfigException("ConfigError:  --startUrl requires a value");
      }
      start_url = args[i + 1];
      has_start_url = true;
      ++i;  // Skip the value
    } else if (arg == "--windowId") {
      if (i + 1 >= args.size()) {
        throw InvalidConfigException("ConfigError:  --windowId requires a value");
      }
      try {
        long window_id_val = std::stol(args[i + 1]);
        window_id = static_cast<uint32_t>(window_id_val);
        has_window_id = true;
        ++i;  // Skip the value
      } catch (const std::invalid_argument&) {
        throw InvalidConfigException("ConfigError:  --windowId value is not a valid integer");
      } catch (const std::out_of_range&) {
        throw InvalidConfigException("ConfigError:  --windowId value is out of range");
      }
    } else {
      throw InvalidConfigException(
          "ConfigError: Unknown argument '" + arg + "'"
      );
    }
  }

  // PHASE 2: Validate all parsed values AFTER iteration completes
  
  // Verify all required arguments are present
  if (!has_ipc_port) {
    throw InvalidConfigException("ConfigError:  Missing required argument: --ipcPort");
  }
  if (!has_session_token) {
    throw InvalidConfigException("ConfigError:  Missing required argument: --sessionToken");
  }
  if (!has_start_url) {
    throw InvalidConfigException("ConfigError:  Missing required argument: --startUrl");
  }
  if (!has_window_id) {
    throw InvalidConfigException("ConfigError:  Missing required argument: --windowId");
  }

  // Validate ipcPort range [1, 65535]
  if (ipc_port < 1 || ipc_port > 65535) {
    throw InvalidConfigException("ConfigError:  --ipcPort must be in range [1, 65535]");
  }

  // Validate windowId range [0, UINT32_MAX]
  if (window_id < 0 || window_id > UINT32_MAX) {
    throw InvalidConfigException("ConfigError:  --windowId is out of valid range");
  }

  // Validate startUrl must start with https://
  if (start_url.rfind("https://", 0) != 0) {
    throw InvalidConfigException(
        "ConfigError: --startUrl must start with https://"
    );
  }

  return AppConfig(ipc_port, session_token, start_url, window_id);
}

uint16_t AppConfig::GetIpcPort() const {
  return ipc_port_;
}

const std::string& AppConfig::GetSessionToken() const {
  return session_token_;
}

const std::string& AppConfig::GetStartUrl() const {
  return start_url_;
}

uint32_t AppConfig::GetWindowId() const {
  return window_id_;
}

}  // namespace core
}  // namespace cef_ui
